OpenTracing support

tcolgate opened the pull request:
### Description

Support tracing initial request arrival and outbound backend calls.
This adds hooks into the middleware to inject span tracing. It uses the OpenTracing API, and this
code uses the Jaeger Tracer. If this is something people want, I'll look at adding other opentracing compliant tracer support (e.g. zipkin).

Obviously needs lots of work to make this acceptable, but I wanted to see if people were interested.

Fixes #771, #390

ldez commented:
Huge !

@tcolgate Could you change the base branch to master?

We only accept bug fixes in 1.4.

tcolgate commented:
It wasn't immediately clear to me how dependencies are being handled (the glide scripts proved a little painful. What's the correct process to introduce a dependency?

ldez commented:
@tcolgate could you split your commit in 2 commits:
- first the code
- second the vendor + glide

If you authorize me, I can manage the glide part.

tcolgate commented:
@ldez  I have split the PR, and dropped the inclusion of the actual deps.
Happy for you to manage the glide part.

I'm thinking:
- have a tracing.tracer string settting to pick which tracer to use (perhaps default to jaeger
- have tracer specific setting under tracing.jaeger.... and tracing.zipkin.... 
- move the actual tracing loging int middleware/tracer/
- Potentially hook logging and metrics up to traefik's internal logging/metrics.

In terms of more functionality, it could be possible to add tracing awareness to other middlewares

The results are already quite pleasing, you get the initial ingress trace, and if you have, e.g., retry enabled, you get a trace for each attempt. Plus trace ids get allocated so should be propagated to downstream apps.

tcolgate commented:
Here's an example of a jaeger trace of a request to a traefik with 3 retries enabled to a bad backend.
![tracing-example](https://user-images.githubusercontent.com/76019/31379717-04306136-ada7-11e7-8654-8ae90ea39b7c.png)

martinbaillie commented:
Looking good @tcolgate thanks for getting started on OpenTracing! I swapped my production gateway from Zuul to Traefik recently. Tracing is the one thing I lost in that switch and has been on my list to add to the Traefik codebase.

I'm happy to help out any way I can. Also I _do_ use Zipkin currently and think abstracting from tracer is a good idea, so let me know if I can be of assistance there.

tcolgate commented:
Could you provide examples of spans from zuul, or point me at their code
Idn not quite sure what tags to include right now

On Tue, 10 Oct 2017, 23:34 Martin Baillie, <notifications@github.com> wrote:

> Looking good @tcolgate <https://github.com/tcolgate> thanks for getting
> started on OpenTracing! I swapped my production gateway from Zuul to
> Traefik recently. Tracing is the one thing I lost in that switch and has
> been on my list to add to the Traefik codebase.
>
> I'm happy to help out any way I can. Also I *do* use Zipkin currently and
> think abstracting from tracer is a good idea, so let me know if I can be of
> assistance there.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/containous/traefik/pull/2237#issuecomment-335628069>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEo85LlAnbKSxwx0LfGHLmRrMc4taEjks5sq_D6gaJpZM4PzjI0>
> .
>

martinbaillie commented:
Here's Spring Cloud's take on OpenTracing:
https://github.com/spring-cloud/spring-cloud-sleuth
Their version of Zuul is here:
https://github.com/spring-cloud/spring-cloud-netflix/tree/master/spring-cloud-netflix-core
but
I believe it'll ultimately use Sleuth as part of client-side load balancing
through Feign/Ribbon, like (here)[
https://github.com/spring-cloud/spring-cloud-sleuth/blob/master/spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/feign/TraceFeignClient.java
]

On Wed, 11 Oct 2017 at 16:26 Tristan Colgate-McFarlane <
notifications@github.com> wrote:

> Could you provide examples of spans from zuul, or point me at their code
> Idn not quite sure what tags to include right now
>
> On Tue, 10 Oct 2017, 23:34 Martin Baillie, <notifications@github.com>
> wrote:
>
> > Looking good @tcolgate <https://github.com/tcolgate> thanks for getting
> > started on OpenTracing! I swapped my production gateway from Zuul to
> > Traefik recently. Tracing is the one thing I lost in that switch and has
> > been on my list to add to the Traefik codebase.
> >
> > I'm happy to help out any way I can. Also I *do* use Zipkin currently and
> > think abstracting from tracer is a good idea, so let me know if I can be
> of
> > assistance there.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/containous/traefik/pull/2237#issuecomment-335628069
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AAEo85LlAnbKSxwx0LfGHLmRrMc4taEjks5sq_D6gaJpZM4PzjI0
> >
> > .
> >
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/containous/traefik/pull/2237#issuecomment-335686962>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAldbL2nGmGYZLmUXKMGFLKUj5Uh2xWTks5srFF9gaJpZM4PzjI0>
> .
>

tcolgate commented:
Started work on how to config different tracers, this code is untested right now due to time, but will work on it further.
  Once jaeger+zipkin are working, I'm going to move the tracing code into the middleware directly and remove the use of the httplib middleware in favour of a custom one

tcolgate commented:
- Rebased on master
- Added some utilities to let other things use tracing
- Added tracing to the retry middleware.

Retry also includes an example of adding log events to a trace.

tcolgate commented:
@martinbaillie  the zipzkin tracing is sending stuff to a server, but for some reason, only the first span is showing up, I'm not sure what could be causing that.

tcolgate commented:
Main middlewars are now instrumented.
Interestingly both jaeger and zipkin see the same shape spans from the zipkin export. Either the opentracing zipkin support is broken, or I'm not setting the tracing up properly.
Jaeger export is fine and working as expected.

tcolgate commented:
@idez I'm struggling to track down the zipkin issue. Jaeger funcionality seems good right now. I'm not sure what the best route forward is. Maybe ship v1 without zipkin?

tcolgate commented:
I'm about 80% sure zipkin issues are due to me doing something wrong.
I've checked out opentracing, zipkin-opentracing-go, and thrift, at the same commits as we are vendoring in this PR. The example app show proper, nested traces. Will keep digging

tcolgate commented:
@martinbaillie @ldez Fixed the zipkin issue. I have a few tidyup things to do around config. But things are working. Remaining:
- Allow configuration of sampling setting for zipkin
- Some test (opentracing has a mock, so that much should be easy enough)
- Work out if config handling is inline with the projects preferences.

tcolgate commented:
![zipkin](https://user-images.githubusercontent.com/76019/31686342-c83c65d0-b37d-11e7-9eca-ed10bf08c841.png)

mmatur commented on middlewares/tracing/tracing.go:
```diff
@@ -0,0 +1,204 @@
+
+// NewSpanMiddleware creates a new middleware wraps a span around subsequent
+// middleware
+func (t *Tracing) NewSpanMiddleware(name string) negroni.Handler {
```
`NewSpanMiddleware` seems to not be used. What is his purpose ?

tcolgate replied:
I intended it as the main means for adding tracing to new middlewares, however that didn't pan out. I'll remove it.

ldez commented:
@tcolgate So that your PR is integrated in the 1.5, it is imperative that it is ready by Tuesday 21 December at the latest. :rocket:

ldez commented:
Sorry November

tcolgate commented:
I've started tidying this up, but don't have a tonne of time. I'm hoping I shokd get it done in the next couple of weeks though.

tcolgate commented:
This is probably going to have to wait until the new year now. Sorry about that.

errm commented:
No problem, thanks for your work so far @tcolgate 👍

mmatur commented:
@tcolgate Due to you lack of time, we have decided to carrying your PR in #2587. 

Thanks for your hard work :clap: .
